Enhance: get splitter on right side of editor working + translatable test

SlySven opened the pull request:
Restructure the `trigger_editor.ui` form so that a single splitter on right hand side is used for all widgets on that side which makes it possible to resize the various elements in a more flexible way than previously.

Make item help messages be translatable and ensure ones at top of `dlgTriggerEditor.cpp` are used (rather than others buried in the add item of each type) - and allows the space for them to be expanded if needed.

Make column headers in search results `treeWidget` be translatable.

Revised the code that reported items being enabled/disabled in the editor so that if there is an error the item is always left disabled instead, it also reformulates the texts so that they can be translated as a sentence, with less fragments of text components.

Simplifies the layout of the `triggers_main_area.ui` and arranges so that it can expand as necessary for other GUI languages.

Improves sound file selection on trigger UI - including presetting file selector to previously chosen file or profile home directory (instead of user home dir) if none and adding a tool button to clear the setting - now the use of cancel in the file selection dialog does indeed cancel the process of *setting* a file rather than clearing a previous setting - which is not what the user would have expected...

Improve colourising fore/background colour selection UI - including greying out colour buttons when disabled.

Remove a load of casts from `enums` relating to the type of an editor item to `int`s by making passed arguments to some methods to explicitly be those enums {which, after all, is one purpose of using enums in the first place!}

Revise the `timer_main_area.ui` with less fixed sizes that won't expand as needed.

Demote several methods from being `SLOTS` in `dlgTriggerEditor` class:
* `(void) dlgTriggerEditor::slot_action_toggle_active()` ==> `dlgTriggerEditor::activeToggle_action()`
* `(void) dlgTriggerEditor::slot_alias_toggle_active()` ==> `dlgTriggerEditor::activeToggle_alias()`
* `(void) dlgTriggerEditor::slot_key_toggle_active()` ==> `dlgTriggerEditor::activeToggle_key()`
* `(void) dlgTriggerEditor::slot_script_toggle_active()` ==> `dlgTriggerEditor::activeToggle_script()`
* `(void) dlgTriggerEditor::slot_timer_toggle_active()` ==> `dlgTriggerEditor::activeToggle_timer()`
* `(void) dlgTriggerEditor::slot_trigger_toggle_active()` ==> `dlgTriggerEditor::activeToggle_trigger()`
* `(void) dlgTriggerEditor::slot_deleteAlias()` ==> `dlgTriggerEditor::delete_alias()`
* `(void) dlgTriggerEditor::slot_deleteAction()` ==> `dlgTriggerEditor::delete_action()`
* `(void) dlgTriggerEditor::slot_deleteKey()` ==> `dlgTriggerEditor::delete_key()`
* `(void) dlgTriggerEditor::slot_deleteScript()` ==> `dlgTriggerEditor::delete_script()`
* `(void) dlgTriggerEditor::slot_deleteTimer()` ==> `dlgTriggerEditor::delete_timer()`
* `(void) dlgTriggerEditor::slot_deleteTrigger()` ==> `dlgTriggerEditor::delete_trigger()`
* `(void) dlgTriggerEditor::slot_deleteVar()` ==> `dlgTriggerEditor::delete_variable()`
* `(void) dlgTriggerEditor::grab_key_callback()` ==> `dlgTriggerEditor::key_grab_callback()`

Also removed some that were wrappers around short code fragments that could be put directly in the single place of use:
* `(void) dlgTriggerEditor::slot_addTrigger()`
* `(void) dlgTriggerEditor::slot_addTriggerGroup()`
* `(void) dlgTriggerEditor::slot_addTimer()`
* `(void) dlgTriggerEditor::slot_addTimerGroup()`
* `(void) dlgTriggerEditor::slot_addAlias()`
* `(void) dlgTriggerEditor::slot_addAliasGroup()`
* `(void) dlgTriggerEditor::slot_addScript()`
* `(void) dlgTriggerEditor::slot_addScriptGroup()`
* `(void) dlgTriggerEditor::slot_addAction()`
* `(void) dlgTriggerEditor::slot_addActionGroup()`
* `(void) dlgTriggerEditor::slot_addKey()`
* `(void) dlgTriggerEditor::slot_addKeyGroup()`
* `(void) dlgTriggerEditor::slot_addVar()`
* `(void) dlgTriggerEditor::slot_addVarGroup()`

Renames some `.ui` form components that are manipulated by the code in the `dlgTriggerEdit` class so that the element concerned is more clearly identified and consistent with other names, e.g.:
`soundTrigger` ==> `groupBox_soundTrigger`

Correct the tool-tip text for the `spinBox_stay_open` from:
"Within how many lines must all condition be true to fire the trigger?"
to:
"How many more line after the one that fired the trigger should be passed to children?"
as the former text was an erroneous duplicate of the tool-tip for the delta value for multi-line or AND triggers.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
This is a bit of an unwieldy beast :slightly_frowning_face: of a PR but it distils several interrelated fixes to the editor which should improve the UI in that area.

:earth_africa: **I18n** It will put up some more texts that need translation that were previously not flagged for it.

It (*finally*) implements fixes I was originally contemplating for #436 a PR that got abandoned a couple of years ago.

xekon commented:
On windows I seem to be unable to resize the right side at all, there does not appear to be any drag handle.

vadi2 commented:
I'd say as far down as one would be nice!

SlySven commented:
> I am able to resize it, but it wont let me make it any smaller than 8 patterns visible, would be nice if I could shrink it down as far as 2 patterns visible

The limiting factor may be the post-trigger run-on (?)/sound on trigger options - it should be preventing those from being squashed and their height should be what sets the *minimum* height for that group (middle) of widgets.  Pictures would be useful if that is NOT the case, @xekon .

xekon commented:
Ah you might be right, it looks like those controls use a bit more space now. It does look like you worked to improve the look of the form as well which is nice. If there is anything you can do to make the controls a little more compact it would allow us to collapse it upward a bit farther.

That way when we are working in the code window we can have as much space as possible.

here is a picture that shows the before and after difference between the two, I cannot collapse either beyond this:

![old-new-comparison](https://user-images.githubusercontent.com/106644/55986687-d7db6a00-5c54-11e9-899f-eee8a341e52c.png)

xekon commented:
For instance t he Fg Color and Bg Color buttons could be side by side, instead of stacked vertically.

vadi2 commented:
I agree, we should allow controls to go beyond their mibimumt size. The user knows better how small they want to go.

SlySven commented:
Some widgets - `QLabel`s I think are difficult - to get to adjust themselves automagically and still have a reasonable size.  As for the trigger colourise option that is in the **bottom** not the **middle** area at present - so has no effect on the minimum height that the middle part can have, I *think*... :slightly_frowning_face:

Let me see if I can rearrange things better - and - are there any objections to putting the right and bottom controls in different locations?

xekon commented:
This picture better illustrates why I was saying to make the color buttons smaller and place them side by side:

![2019-04-11_16-00](https://user-images.githubusercontent.com/106644/55998412-f3ee0400-5c72-11e9-88c5-c2eb4ac15e02.png)

SlySven commented:
Ah, in that case they are the limiting factor that determines the height of the bottom part - but what about if the width is less so that, say, the text in the `match all` box is wrapped once or twice...

... also I am not entirely comfortable about using `fg color` and `bg color` - the first part of each is not a good term to use and there is no guarantees that translations in other languages can be anything like as short.

I am still playing with the layout at the moment - I had what I thought was a better one but I am getting weird code model errors from the built-in Qt one (not the clang one, I normally disable that as it was feasting on memory - I was getting it swallowing over a couple of GB in the past but I am just gonna switch it back in instead to see if it makes more sense!)

SlySven commented:
This is a bit better I suppose - it is at pretty much the minimum size it will resize down to in that state on my current OS/DE setttings - and it is showing a shade under 4 lines of items:
![Screenshot_20190412_022951](https://user-images.githubusercontent.com/6163092/56006925-b1f2ab80-5cce-11e9-9b32-e9d6cbf456d0.png)

Note that I have moved the sound on trigger option in the bottom row and moved the other spin box control into the place it came from - which reduces the minimum height the middle area can go down to - but with two lines of items inside it there is less benefit to placing the two trigger coloriser buttons side-by-side - although when I put them on top of each other the effects seemed worse
![Screenshot_20190412_030932](https://user-images.githubusercontent.com/6163092/56007389-70fb9680-5cd0-11e9-95b5-749da98489f5.png)

vadi2 commented:
![image](https://user-images.githubusercontent.com/110988/56029493-c197e180-5d1a-11e9-87f1-0812b0a86788.png)

Can we make all of the red area be a box that has no minimum height, so we can cut off as much as we want?

SlySven commented:
> Can we make all of the red area be a box that has no minimum height, so we can cut off as much as we want?

I don't think so - we are using layouts and the whole purpose of them is to make the best use of the space available depending on what is placed inside them.  Part of the black-magic I seem to be doing is to try and guide that so that it does that without causing things to squash up so much that controls become ridiculous with texts going outside of the view port (unless it is a widget with automatic scrollbars or elisi...) or anonymous buttons that squash down to a thin strip.

Getting those things to actually happen in a *controlled* manner *might* be possible - but which widgets can get clobbered?  I suppose it would be feasible to allow for elision in `match all` groupBox - which could gain us a few pixels in the vertical direction, but I am not sure that the text display part (of that) has that property setting...

In the example shown - don't forget you can turn off the `TConsole` instance in the bottom right (is the only one in the profile that has the `TConsole::mType` set to `TConsole::ErrorConsole`) - which will double up the space available for script editing in that minimum case.  I was just using it to demonstrate the worst case with the minimum space usage.

SlySven commented:
I have a revision but since I switched to Linux it is not comparing like-for-like - I converted the two `QGroupbox`s now on the bottom left of the `triggers_main_area` to be plain `QCheckbox`s with the former text as tooltips...

![Screenshot_20190412_223104](https://user-images.githubusercontent.com/6163092/56067554-7cec6480-5d73-11e9-923e-cc1ffacc8995.png)

This (and also changing `begin of line sububstring` to `start of line` so that the QComboBox for the item type is smaller) allows the minimum width to be reduced but I'd need to try it back on FreeBSD to see whether it makes an improvement for the minimum height.  As it is I think it will be minimal as the limiting factor (now, again) the sound on trigger component which has three *lines* of details:
* the groupbox checkbox & text
* the button to change the file
* the display of the file name and the clear button

Whilst we had hidden the file name display before, it is user unfriendly (IMVHO) to do that as it makes it difficult to check what the setting is (although, at least now it won't get nuked if you click on the button to bring up the file selector and then cancel out like it did **before** this PR)...

vadi2 commented:
I'm not seeing this to be much of an upgrade :confused: 

It's trying to solve a problem we didn't have - making the trigger patterns area bigger.

Instead it adds problems: triggers pattern area is bigger by default (6 lines vs 5), it shuffles controls around which'll frustrate people who are used to them, it increases the font size so it doesn't look as nice anymore, and it renames some controls to add more confusion.

There definitely are improvements in the code that makes this work, improvements to the sound selector and maybe others, but the problems it adds don't outweight the problems its fixing...

vadi2 commented:
I'm having a look at it as well, trying to see how can we make the trigger patterns area smaller...

vadi2 commented:
I've had to take the actual scroller out to make it work, but here's what I mean with a patterns area that can be made as small as desired:

![Peek 2019-04-14 11-18](https://user-images.githubusercontent.com/110988/56090783-30a24100-5ea7-11e9-907d-2e2ab8cf5a18.gif)

Taking a trans-Atlantic flight now so attaching my WIP work.

[triggers_main_area_working_sans_patterns.zip](https://github.com/Mudlet/Mudlet/files/3077208/triggers_main_area_working_sans_patterns.zip)

SlySven commented:
Some observations:

> ... it shuffles controls around which'll frustrate people who are used to them ...

That is being done to try and optimise the layout so that it can be made as compact as possible without loosing functionality and also allow for the texts in controls to grow as they likely will for other languages - I found this interesting page on the MediaWiki site and whilst it is not directly relevant to our situation it does list some issues we will likely encounter and it does specifically warn: **[Messages are often longer than you think!](https://www.mediawiki.org/wiki/Localisation#Messages_are_often_longer_than_you_think!)**

> It's trying to solve a problem we didn't have - making the trigger patterns area bigger.

It is trying to solve a problem that the prior code did not *allow the trigger patterns area to be made larger* - which **is** an issue e.g. #1336
.
> ... it increases the font size so it doesn't look as nice...

If the fonts are ~~getting bigger~~ changing size than that means they were being overridden from the defaults that the user is seeing from their Desktop Environment. That is at least *unwise* (IMNSHO) because those will **not** track changes that the user or their DE choices may have set. {*By the same token I deliberately make sure that any forms that I work on do not apply font settings - some past developers have, previously (and hopefully* unintentionally *), left some very OS specific fonts which are not available on every OS we support!*} In the very few cases where a change of font size is needed I resort to `<big>`...`</big>` and `<small>`...`</small>` tags in the rich-text which will still work with whatever is specified externally.

vadi2 commented:
I don't know. It's a hard sell to tell someone "we shuffled things around. Trust me, it's better. Sorry you've gotten used to it."

SlySven commented:
How about "we've shuffled things about to make the best use of the space because we need some extra for some languages other than English/Chinese"?

I have tried to replicate the effect you have in that dummy `.ui` file. The only reason you can collapse the middle area over the widgets that remain after you have remove the scroll area where the trigger items are normally held is because those remaining widgets are not included in a `QLayout` {the base class for geometry managers} derived class so are not considered in the arrangement of widgets to be shown. As such there is ***nothing*** to manage/adjust their size and placement - so there is nothing that makes them *fit* in among the other widgets - and thus they will not be adjusted if the form is changed in size - and they will be completely hidden and inaccessible when the trigger items are put back as the `(QVBoxLayout) lay1` in the `dlgTriggerEditor` constructor will grab all the space inside the middle area.

So, unfortunately, I do not think that is a workable approach. :frowning_face:

SlySven commented:
However - we already have one control to show/hide the search results so I think I can add a similar control on the other side to show/hide the controls to the right and below the trigger items - which should make it possible to shrink the trigger_main_area up to just show the **name** and **command** and 1 to 2 trigger items and a control to show/hide the rest of the controls...

SlySven commented:
What about this then - with a control to Show/Hide things:

![Screenshot_20190416_031028](https://user-images.githubusercontent.com/6163092/56177109-6d7c4e00-5ff5-11e9-8453-1ab9e3d28139.png)

which can be made smaller by clicking on that control:
![Screenshot_20190416_030926](https://user-images.githubusercontent.com/6163092/56177133-7ff68780-5ff5-11e9-9ae3-d3c3139e5b8b.png)

I would be tempted to put it in the top row between the **Name:** and **Command:** field...

Kebap commented:
I would expect bigger layout changes in Mudlet 4.0 maybe but not in 3.2x

The idea to hide controls seems interesting, but the current position next to the trigger texts takes up too much screen space by itself

vadi2 commented:
Russian text is one of the longer ones out there and it does not require a redesign. Everything fits.

![Selection_420](https://user-images.githubusercontent.com/110988/56198492-09fc1b80-6009-11e9-9c28-8350ee826a53.png)

I'm not blindly against a redesign, but I am against one that wasn't warranted - if we need to add a splitter, I don't see why can't we without moving everything else around!

SlySven commented:
And how big is the whole dialogue on that Russian version? What does it look like if you adjust it to the minimum size (or if you were using Mudlet on a device with a screen of only, say 1024x768 i.e. XGA)?

The trouble is that the triggers view has the most *knobs* on it. With a minimum size to feature elements (Qt *global strut* size) getting a form that is still usable when everything is at it's minimum size yet allow some elements to expand to take advantage of the extra space to show more information does require a certain amount of juggling.  In the past there were some elements in the `triggers_main_area` that were of fixed size which I have now make expandable but it is never going to be perfect for everybody all of the time.

BTW The splitter was already there but it only handled the lua script area and the errors TConsole - I've extended it to also handle the xxxx_main_areas and the system message area - which allows for more end-user control but needed revision to those xxxx_main_areas to make them handle having the space available to them be adjustable...

xekon commented:
@slysven where is your latest build of this? I would like to try it out.

vadi2 commented:
See https://github.com/Mudlet/Mudlet/pull/2507#issuecomment-481928767

SlySven commented:
@xekon try the latest commit - I've redone the hide the controls button so that it is just the icon with no text (but with an explanatory tool-tip) on the top line between the **Name:** and **Command:** entry items... you may just have to wait about for the CI to NOT be:

**THESE ARE NOT THE MOST UP TO DATE ONES - DO NOT USE**
linux: https://transfer.sh/1EHF0/Mudlet-3.18.0-testing-PR2507-ffafe4a-linux-x64.AppImage.tar
osx: https://transfer.sh/7Xo07/Mudlet-3.18.0-testing-PR2507-ffafe4a.dmg
windows: https://ci.appveyor.com/api/buildjobs/s8h9fjilj1qh21e4/artifacts/src%2Fmudlet.zip

I will delete this listing later on today but I will be away for a bit and don't want you to grab the wrong ones (these!)

xekon commented:
I just tried the Windows Build! That is a pretty cool control! I wonder if that control could be made to work automatically. Depending on if your trying to collapse the area smaller than possible. Then when you drag the area larger the extra controls could automatically show. Then it would operate similarly to how Vadi posted here: https://github.com/Mudlet/Mudlet/pull/2507#issuecomment-482937146

SlySven commented:
:thought: It *might* be possible by measuring the span occupied by the `triggers_main_area` widget as a whole and using that to trigger the control - however then you have to set a threshold as to when it would happen and include some hysteresis so it does not oscillate between the two states.  You also have to worry about how the required threshold might depend on individual users' DE (font, font size, language {different languages will probably take different amounts of space}) and getting that right for everyone could be tricky...

xekon commented:
Ahh I see... hmmm

SlySven commented:
These **are** the binary installers you are looking for containing 6f5e9ec:

linux: https://transfer.sh/YVft3/Mudlet-3.19.0-testing-PR2507-6f5e9ec-linux-x64.AppImage.tar
osx: https://transfer.sh/IHWsL/Mudlet-3.19.0-testing-PR2507-6f5e9ec.dmg
windows: https://ci.appveyor.com/api/buildjobs/qvj5i6meyqx0hgg7/artifacts/src%2Fmudlet.zip

They do try to show/hide those controls automagically now (the threshold is more than 350 pixel high to show and less then 300 to hide them).

vadi2 commented:
I think this auto hider/expander works great. Can we put the controls back to how they were so we're not destroying our users muscle memory?

xekon commented:
I just Tried the auto hider/expander as well. I love it!

SlySven commented:
> ... Can we put the controls back to how they were so we're not destroying our users muscle memory?

I've tried to and found It rather compromises the layout - with the controls showing - as the `play sound` control benefits from extra horizontal space but that can only be got on the previous placement by reducing the width for the items' display.

That control also requires more vertical space than the (now) single spinBox for the `AND / Multi-line (delta)` control with the result that the height threshold where the extra controls ***have*** to be hidden is a larger value than the layout I am proposing - which is unhelpful on a smaller screen device (say 1024 x 768) like my mobile (laptop) development platform...

SlySven commented:
Some screen shots with the dialogue shrunk to the **minimum** it allows - to show the differences:

**My proposal** - with no errors console, **controls hidden** = (666 x **334**, ***3*** item rows):
![Screenshot_20190420_224122](https://user-images.githubusercontent.com/6163092/56462581-6ff7f280-63bd-11e9-80e0-431aa15fa51b.png)

Reversion to original control locations = (666 x **439**, ***7*** item rows):
![Screenshot_20190420_222102](https://user-images.githubusercontent.com/6163092/56462522-64f09280-63bc-11e9-84bf-684705e3ad97.png)

**My proposal** - with no errors console, controls shown = (980 x **498**, ***5*** item rows):
![Screenshot_20190420_222920](https://user-images.githubusercontent.com/6163092/56462524-6f129100-63bc-11e9-867c-b539da1f2ad0.png)

Reversion to original control locations = (1050 x **510**, ***6*** item rows):
![Screenshot_20190420_225127](https://user-images.githubusercontent.com/6163092/56462656-ddf0e980-63be-11e9-89a1-fb11e5b2b772.png)

Note that the extra `line delta` label in the reversion is probably increasing the minimum width in the second comparison.

TBH It *might* be possible to squeeze the layout a bit more with the controls in the original location but it will need more investigation and tweaking of various minimum height settings in some places and their interaction is not always obvious and not really something I want to pursue at this point; I just want to get this PR in so that I can tick it off my *ToDo* list... :stuck_out_tongue_closed_eyes:

vadi2 commented:
How would you feel if `git` changed its syntax for X reasons? It'd be pretty annoying - so we have to look into how we can do this with minimal disruption.

xekon commented:
So if we revert the controls to the original position, then the automagical hide of controls would happen at how many pixels? currently its at 300-350 which feels pretty good.

SlySven commented:
Okay - back on FreeBSD and with some tweaks (colorizer highlight buttons layout out horizontally) and the original control position layout (although the colorizer control is now grouped vertically with the other controls on the right and the other three at the bottom are laid out in a sub-group that is then grouped with the scrolling-area that contains the trigger items). Anyhow, the minimum size for the view on the same system as the first screen-shots:

Reversion to original control locations, with controls shown = (1062 x 503, 4 item rows):
![Screenshot_20190424_001801](https://user-images.githubusercontent.com/6163092/56622069-9371a600-6626-11e9-849f-841ed50326d7.png)

Reversion to original control locations, with controls hidden = (670 x 401, 5 item rows):
![Screenshot_20190424_001721](https://user-images.githubusercontent.com/6163092/56622070-940a3c80-6626-11e9-8cf8-15793dbdbf6f.png)

With the error window shown and the very first image again above it looks like:
![original-image](https://user-images.githubusercontent.com/6163092/56006925-b1f2ab80-5cce-11e9-9b32-e9d6cbf456d0.png)
![current-image](https://user-images.githubusercontent.com/6163092/56622455-219a5c00-6628-11e9-81a7-20b3388b5fcc.png)

SlySven commented on src/dlgTriggerEditor.cpp:
```diff
@@ -592,17 +610,14 @@ dlgTriggerEditor::dlgTriggerEditor(Host* pH)
+    labelList << tr("Type", "Heading for the first column of the search results display treewidget")
+              << tr("Name", "Heading for the second column of the search results display treewidget")
+              << tr("Where", "Heading for the third column of the search results display treewidget")
+              << tr("What", "Heading for the four column of the search results display treewidget");
```
Agreed - will fix.

SlySven commented on src/dlgTriggerEditor.cpp:
```diff
@@ -2842,21 +2777,23 @@ void dlgTriggerEditor::slot_timer_toggle_active()
-        iconError.addPixmap(QPixmap(QStringLiteral(":/icons/tools-report-bug.png")), QIcon::Normal, QIcon::Off);
-        pItem->setIcon(0, iconError);
+        pT->setIsActive(false);
+        showError(tr(R"(Unable to activate (and automatically deactivating) an offset timer or timer or the part of a module "%1" that contains them; reason: %2.)").arg(pT->getName(), pT->getError()));
```
> Since we're unable to activate it, it means that it's only other possible state is that it is deactivated.

Not exactly - it is, what I have previously started to use: *bugged-out* - i.e. in an error state - which, whilst it effectively the same as disabled in that it does not do anything - means it cannot be activated without user action however many times they click `activate`.

The text in this area, *importantly* also covers the case which we haven't really dwelt on before - of a ***module*** - where the brown folder is merely a representation of (in this case, the timer type) elements of that module - the significance of that is **it is a dummy *and cannot be edited!*** It IS possible to disable all of that type of elements in the module from the editor - but anything else is not effective - anything the user enters in any of the fields is forgotten when they switch to another item (and changing the name is not possible because it is set from the module name and also cannot be changed - and given that there is likely to be identical folders in other views that is just as well,..!)

SlySven replied:
Also, the code previous to this commit would report that (in the case of said bugging-out) it was not possible to **deactivate** the item for whatever reason - which is confusingly rong! This text also handles the situation where it *was* activate but the user has since saved some new, bugged code - this PR means that it now forces the item to be deactivated (and is saying so) - and they will need to reactivate it once any bugs have been fixed.  This might sound a bit less friendly but it is also less confusing - if it is broken the user must switch it back on after it is fixed - of course the double-click on the item in the treewidget will do both the save (and compile - which will clear the error state) and then the activation...

vadi2 replied:
Technically true, but practically confusing. I'd still opt for the simpler version.

SlySven commented on src/dlgTriggerEditor.cpp:
```diff
@@ -592,17 +610,14 @@ dlgTriggerEditor::dlgTriggerEditor(Host* pH)
-              << "Name"
-              << "Where"
-              << "What";
+    labelList << tr("Type", "Heading for the first column of the search results display treewidget")
```
Perhaps it would be more precise to say: "Translators may not be coders?" :stuck_out_tongue_winking_eye: 

However, you probably do have a point there - OTOH I was using the term we use for the widget precisely because that is what we (well someone deep in the Mudlet past) decided to call it. I will cull it though.

SlySven commented on src/dlgTriggerEditor.cpp:
```diff
@@ -7888,18 +7859,41 @@ void dlgTriggerEditor::slot_colorizeTriggerSetBgColor()
+                                                    mpTriggersMainArea->lineEdit_soundFile->text().isEmpty()
+                                                    ? mudlet::getMudletPath(mudlet::profileHomePath, mpHost->getName())
+                                                    : mpTriggersMainArea->lineEdit_soundFile->text(),
+                                                    tr("Audio files(*.aac *.mp3 *.mp4a *.oga *.ogg *.pcm *.wav *.wma);;"
```
`QMediaPlayer` is quite adept at playing things and it possibly might depend on the underlying system libraries that it can dynamically link to - so I do not have any reason to not believe Mudlet can handle them - it is just that adding the filters makes it easier if the user is looking for a particular file if they can hide files that are definitely NOT audio files...

SlySven commented:
>trigger patterns open in giant view by default (23 lines for me), I guess that's because my script editor occupies most of my screen

Yep, it is making use of the space available - as intended! :grinning: If you resize the Lua script window there will be less items shown and the editor will, I think, remember that adjustment when the view is switched.

>  "only pass matches" and "match all" lost tooltips

I'll have a look for them... :mag_right:

vadi2 commented:
It doesn't remember that adjustment when you restart Mudlet, though. Also most people have more code in a trigger than patterns - it makes more sense to show more code by default I believe.

SlySven commented:
>>> trigger patterns open in giant view by default (23 lines for me), I guess that's because my script editor occupies most of my screen
>>>
>> Yep, it is making use of the space available - as intended! grinning If you resize the Lua script window there will be less items shown and the editor will, I think, remember that adjustment when the view is switched.
>>
> It doesn't remember that adjustment when you restart Mudlet, though. Also most people have more code in a trigger than patterns - it makes more sense to show more code by default I believe.

That would require saving / restoring the state of the splitter concerned {with `(QByteArray) QSplitter::saveState() const` and `(bool) QSplitter::restoreState(const QByteArray &state)`} or given that there is a second (vertical) one **both** of them between sessions as part of **profile specific layout** data. However we do not do ***that*** yet: note that we save the editor's size and position but we haven't even got that right: we store those characteristics in the main Mudlet `QSetting`s file but we never bother to note which profile the settings were from so every profile opens it's editor exactly on top of a previously opened profiles one - which is not helpful if multi-playing...

:thought_balloon: It *would* be a useful development to save and restore the splitter position but I think it is something that should be deferred to a future PR as part of a wider fix for remembering the Editor layout...

As for the division of the available space between the list of items and the lua script editor it is being done auto-magically[1] right now - it might be possible fiddle with it but I am not sure at this time how to do that (it would need to handle variations from different OSes/DEs and even the recent revision to auto-show/hide those controls only *just* works {has the right show/hide threshold and hysteresis settings} for both my Linux/FreeBSD OSes setup and they are using the same DE)! :worried: 


[1] [Clarkes' Third Law](https://en.wikipedia.org/wiki/Clarke's_three_laws):
> Any sufficiently advanced technology is indistinguishable from magic.

vadi2 commented:
Sounds difficult. Let's just default to showing more code like we did before instead of more patterns? I think that would be a better user experience.

SlySven commented:
> Sounds difficult. Let's just default to showing more code like we did before instead of more patterns? I think that would be a better user experience.

Precisely, but I'm not sure how to do *that*! :worried:

Within the `triggers_main_area` the widget(s) that contain the trigger item rows are set to expand to use whatever space is available - which I think I have to do to ensure they do but I am not sure whether I can work out how to revise things to do what you want. Given the quantity of the changes in this PR I am looking to getting it with as little further changes as necessary. If the user feedback is that strongly negative then I will try and find out how to fix things...

vadi2 commented:
Could you ask about this in Discord?

SlySven commented:
I've just remembered **[(void) QSplitter::setStretchFactor(int index, int stretch)](https://doc.qt.io/qt-5/qsplitter.html#setStretchFactor)**  - lemme see whether that does anything to help - remember I cannot do a "*Let's just default to showing more code like we did before...*" because we are not doing things like we did ***before***! :astonished:

SlySven commented:
Okay I've just cranked it up from *1* to **3** for the script area (all of them were **1** apart from the system message area which was/is **0**) - on my mobile development platform (Devuan "Ascii" Linux, 1200x800 resolution, but without auto-hiding taskbar) - before (1):

![Screenshot_20190430_171002](https://user-images.githubusercontent.com/6163092/56972545-76c7f780-6b6b-11e9-98d0-370a723a4863.png)

After (3):

![Screenshot_20190430_171153](https://user-images.githubusercontent.com/6163092/56972567-847d7d00-6b6b-11e9-9eb9-d8b54e50735d.png)

TBH That looks good enuff to be going on with... :man_shrugging:

vadi2 commented on src/dlgTriggerEditor.cpp:
```diff
@@ -618,10 +618,10 @@ dlgTriggerEditor::dlgTriggerEditor(Host* pH)
+    labelList << tr("Type", "Heading for the first column of the search results")
+              << tr("Name", "Heading for the second column of the search results")
+              << tr("Where", "Heading for the third column of the search results")
+              << tr("What", "Heading for the forth column of the search results");
```
`fourth`

SlySven replied:
:woman_facepalming: Yeah - forth is something else entirely...

SlySven replied:
Otherwise - is good, yes?

vadi2 replied:
Yup
